h/traversal/contexts.py

```diff
@@ -143,3 +144,38 @@ def organization(self):
         if self.group.organization is not None:
             return OrganizationContext(self.group.organization, self.request)
         return None
+
+
+class GroupUpsertContext(object):
+    """Context for group UPSERT"""
+
+    def __init__(self, group, request):
+        self._request = request
+        self.group = group
+
+    def __acl__(self):
+        """
+        Get the ACL from the group model or set "upsert" for all users in absence of model
+
+        If there is a group model, get the ACL from there. Otherwise, return an
+        ACL that sets the "upsert" permission for authenticated requests that have
+        a real user associated with them via :attr:`h.auth.role.User`.
+
+        The "upsert" permission is an unusual hybrid. It has a different meaning
+        depending on the upsert situation.
+
+        If there is no group associated with the context, the "upsert" permission
+        should be given to all real users such that they may use the UPSERT endpoint
+        to create a new group. However, if there is a group associated with the
+        context, the "upsert" permission is managed by the model. The model only
+        applies "upsert" for the group's creator. This will allow the endpoint to
+        support updating a specific group (model), but only if the request's
+        user should be able to update the group.
+
+        TODO: This and ``GroupContext`` can likely be merged once ``GroupContext``
+        is used more resource-appropriately and returned by :class:`h.traversal.roots.GroupRoot`
+        during traversal
```

Comment:
This seems like a code comment rather than part of a docstring? Docstrings are just meant to document arguments and side effects etc, not talk about code design